server/etcdserver/server.go

```diff
@@ -1067,6 +1069,15 @@ func (s *EtcdServer) applySnapshot(ep *etcdProgress, apply *apply) {
 	ep.confState = apply.snapshot.Metadata.ConfState
 }
 
+func (s *EtcdServer) verifySnapshot(snapshot raftpb.Snapshot) {
+	verify.Verify(func() {
+		cindex := s.consistIndex.ConsistentIndex()
+		if cindex < snapshot.Metadata.Index {
```

Comment:
Do we expect cindex > snapshot.Metadata.Index ? 
I know we can recover from such state, so it's not critical...
but if this is supposed to be defensive invariant for testing, I would assume we should aim for `==`

Reply:
I think it's possible in theory although I am not 100% sure for now, because the entry.Index might be smaller than the cindex. Let's work with an imaginary (& extreme) example, 
1. The leader (S1) partially replicates some log entries at indices `i ~ j`; the leader crashes;
2. Another member (S2) is elected as the new leader, but its local `applied index` & `consistent_index` is still` i-1.` You may challenge this, please look at` figure 8` at `page 8 `in [raft.pdf](https://raft.github.io/raft.pdf);
3. S2 replicates new log entries with index` i` to followers, including S1; S1's local `consistent_index` is `j` (>i),  so which will keep unchanged. 
4. S2 crashes for whatever reason, and S1 is elected as the leader again.  
5. If S1 happens to replicates snapshot to a slower follower of a new member, then the `consistent_index` in the snapshot will be greater than the `snapshot.Metadata.Index`.

Reply:
> 3.  ...S1's local `consistent_index` is `j` (>i),  so which will keep unchanged. But S1's applied index is updated to `i`.

I'm confused in this step. What do you mean by 'consistent_index' ? I believe S1 might have WAL log replicated to `j`, but I assume that the only way to change `consistent_index` is by application. 
When leader S1 crashes and is restarted it's consistent index is populated from latest  snapshot `l<=i` and the only way to advance it is through apply.

Reply:
YES, cindex can only be advanced by applier.

S1's consistent_index and WAL LOG index were advanced to j at step 1. But S2 somehow did not received the message from S1, and it is what the "partially" means at step 1.

Although raft is much simpler than paxos, it is still very subtle and there are some corner cases.

Reply:
You know, etcd applies the entries in a separate goroutine. It means that etcd applies entries [concurrently](https://github.com/etcd-io/etcd/blob/bb63484dcefd7339a2ccbff7ad343a3ef949c381/server/etcdserver/raft.go#L213). So it's possible etcd applies the entries successfully, but fails to replicates the data to some followers. 

So at step1, the leader (S1) might apply the entries successfullly, accordingly the consistent_index is advanced to j. But it failed to replicate the data to S2, accordingly S2's conssitent_index is still i-1.

Reply:
I still don't get it. I understand the Fig.8 story from RAFT. 

But I think that crucial aspect here is that application is always following commitment. 

So: 
 1) The S1 leader must have confirmed j) is replicated to majority of replicas and decided/WAL checkpointed `committedIndex=j`.
      After the restart the S1 follower bootstraps from snapshot & WAL and positions itself (applied/committed/consistent=j). 
  3) In 3, I don't see a code-paths that would allow 'decrease' the applied index from `j->i`. Any replication coming from S2 must be either no-change (if its `index<=committed`) and gets ignored by S1 as we already have it in our WAL log... or it's `index>committed`, so would need to wait for another 'committed' signal before application... and in this scenario `appliedIndex` would only grow as expected.

Reply:
Resolved. PTAL